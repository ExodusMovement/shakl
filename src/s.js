import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
