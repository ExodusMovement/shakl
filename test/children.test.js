import React from 'react'

import { create as r } from 'react-test-renderer'

import s from '../src'

it('passes a single child properly', () => {
  const Foo = s.View({ flex: 1 })

  const foo = r(
    <Foo>
      <Foo />
    </Foo>
  ).toJSON()

  expect(foo.children.length).toBe(1)

  expect(foo).toMatchSnapshot()
})

it('passes multiple children properly', () => {
  const Foo = s.View({})

  const foo = r(
    <Foo>
      <Foo />
      <Foo />
      <Foo />
    </Foo>
  ).toJSON()

  expect(foo.children.length).toBe(3)

  expect(foo).toMatchSnapshot()
})
