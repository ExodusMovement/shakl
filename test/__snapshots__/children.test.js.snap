// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes a single child properly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`passes multiple children properly 1`] = `
<View
  style={Object {}}
>
  <View
    style={Object {}}
  />
  <View
    style={Object {}}
  />
  <View
    style={Object {}}
  />
</View>
`;
